src/cast_core/src/SamplingLoader.cpp

```diff
@@ -0,0 +1,154 @@
+            const auto sampleDocs =
+                std::vector<bsoncxx::document::value>(cursor.begin(), cursor.end());
+
+            if (sampleDocs.empty()) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException("Collection has no documents"));
+            }
+            if (sampleDocs.size() < config->sampleSize) {
+                BOOST_THROW_EXCEPTION(
+                    InvalidConfigurationException("Collection smaller than sample size"));
+            }
+
+            // Maintain an index into the sample that's used for the insert batches so we
+            // insert roughly the same number of copies for each sample.
+            size_t sampleIdx = 0;
+
+            auto batchOfDocs = std::vector<bsoncxx::document::view_or_value>{};
+            auto totalOpCtx = _totalBulkLoad.start();
+            for (size_t batch = 0; batch < config->numBatches; ++batch) {
+                batchOfDocs.clear();
+                for (size_t i = 0; i < config->insertBatchSize; ++i) {
+                    batchOfDocs.push_back(sampleDocs[sampleIdx].view());
+                    sampleIdx = (sampleIdx + 1) % sampleDocs.size();
+                }
+
+                // Now do the insert.
+                auto individualOpCtx = _individualBulkLoad.start();
+                try {
+                    auto result = config->collection.insert_many(std::move(batchOfDocs));
+                    individualOpCtx.success();
+                    totalOpCtx.success();
+                } catch (const mongocxx::operation_exception& x) {
+                    individualOpCtx.failure();
+                    totalOpCtx.failure();
+                    BOOST_THROW_EXCEPTION(x);
+                }
+            }
+            BOOST_LOG_TRIVIAL(info) << "Finished SamplingLoader on thread " << config->threadId;
+        }
+    }
+}
+
+SamplingLoader::SamplingLoader(genny::ActorContext& context, uint thread, size_t totalThreads)
+    : Actor(context),
+      _totalBulkLoad{context.operation("TotalBulkInsert", SamplingLoader::id())},
+      _individualBulkLoad{context.operation("IndividualBulkInsert", SamplingLoader::id())},
+      _client{std::move(
+          context.client(context.get("ClientName").maybe<std::string>().value_or("Default")))},
+      _loop{context, _client, thread, totalThreads, SamplingLoader::id()} {}
+
+class SamplingLoaderProducer : public genny::ActorProducer {
+public:
+    SamplingLoaderProducer(const std::string_view& name) : ActorProducer(name) {}
+    genny::ActorVector produce(genny::ActorContext& context) {
+        if (context["Type"].to<std::string>() != "SamplingLoader") {
+            return {};
+        }
+        genny::ActorVector out;
+        uint totalThreads = context["Threads"].to<int>();
+
+        for (uint i = 0; i < totalThreads; ++i) {
+            out.emplace_back(
+                std::make_unique<genny::actor::SamplingLoader>(context, i, totalThreads));
+        }
+        return out;
+    }
+};
```

Comment:
I'm confused about why the default producer doesn't work for this Actor. Is it because you need the `Threads` config value? If so, you can get that from the `actorContext` (which is accessible from the Phase context iirc). I could be overlooking something here though. Either way a comment might help future maintainers.